xbmc/dbwrappers/dataset.h

```diff
@@ -396,10 +396,30 @@ class Dataset  {
 /* Struct to store an indexMapped field access entry */
   struct FieldIndexMapEntry
   {
-    explicit FieldIndexMapEntry(const char* name) : fieldIndex(~0), strName(name) {}
-    bool operator<(const FieldIndexMapEntry& other) const { return strName < other.strName; }
-    unsigned int fieldIndex;
-    std::string strName;
+    explicit FieldIndexMapEntry(const char* name) : fieldIndex(~0), strName(strdup(name)) {}
+
+    FieldIndexMapEntry(const FieldIndexMapEntry& other)
+      : fieldIndex(other.fieldIndex), strName(strdup(other.strName))
+    {
+    }
+
+    FieldIndexMapEntry& operator=(const FieldIndexMapEntry& other)
+    {
+      fieldIndex = other.fieldIndex;
+      free(strName);
+      strName = strdup(other.strName);
+      return *this;
+    }
+
+    ~FieldIndexMapEntry() { free(strName); }
+
+    bool operator<(const FieldIndexMapEntry& other) const
+    {
+      return strcmp(strName, other.strName) < 0;
+    }
+
+    unsigned int fieldIndex{0};
+    char* strName{nullptr};
```

Comment:
`std::unique_ptr<char>` to avoid manually calling `free`?

Reply:
Not sure this kills the performance gain.

Reply:
A shared_ptr would certainly kill performance (I've done this). I assume unique_ptr compiles down to no overhead on all compilers, but I could be wrong.

Reply:
Sounds reasonable. Will change the code to use a `unique_ptr`.

Reply:
@garbear @lrusak does using unique_ptr here not introduce a malloc/free / new/delete mismatch? `strdup` uses `malloc`, `unique_ptr` uses `delete`, right?

Reply:
Custom Deleter for unique_ptr: https://www.fi.muni.cz/pv264/files/pv264_s09_stdlib.pdf like given as example here? Though I am not sure it's worth it to be honest.

Reply:
I know all this, but is it really worth the effort? Let me just stick with what definitely works, okay?

Reply:
Generally we should use C++ allocation instead of C allocation in a C++ program. As a side effect, we get to use std::unique_ptr for memory safety while avoiding a custom deleter. It would look like:

```c++
// const char* other.strName = "test";

const size_t strLength = strlen(other.strName);

std::unique_ptr<char[]> strName = std::make_unique<char[]>(strLength + 1);

std::copy(other.strName, other.strName + strLength, strName.get());

strName[strLength] = '\0';
```

From a style standpoint, I defintely like your use of strdup and free better. However, from a technical standpoint I think avoiding C allocation is more important.

Reply:
> Or possibly just use std::string again and use the c_str() method to use the strncmp?

@lrusak 

```diff
diff --git a/xbmc/dbwrappers/dataset.h b/xbmc/dbwrappers/dataset.h
index 45f4499905..8d30d9e9b8 100644
--- a/xbmc/dbwrappers/dataset.h
+++ b/xbmc/dbwrappers/dataset.h
@@ -395,7 +395,11 @@ public:
   struct FieldIndexMapEntry
   {
     explicit FieldIndexMapEntry(const char* name) : fieldIndex(~0), strName(name) {}
-    bool operator<(const FieldIndexMapEntry& other) const { return strName < other.strName; }
+    bool operator<(const FieldIndexMapEntry& other) const
+    {
+      // std::string::operator< is too expensive! Use low level function instead.
+      return strncmp(strName.c_str(), other.strName.c_str(), strlen(other.strName.c_str())) < 0;
+    }
     unsigned int fieldIndex;
     std::string strName;
   };
```

Does not really improve performance => `std::string::c_str` now it is.

<img width="1091" alt="Screenshot 2022-08-17 at 08 45 00" src="https://user-images.githubusercontent.com/3226626/185053368-b844bed3-1e7d-40c6-8296-ca436642f120.png">

Reply:
@ksooo Performance > Technical > Style. You do here what works best.